buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchJavaPlugin.java

```diff
@@ -172,26 +163,26 @@ public static void configureCompile(Project project) {
 
                 compileOptions.setEncoding("UTF-8");
                 compileOptions.setIncremental(true);
-
-                // TODO: use native Gradle support for --release when available (cf. https://github.com/gradle/gradle/issues/2510)
-                final JavaVersion targetCompatibilityVersion = JavaVersion.toVersion(compileTask.getTargetCompatibility());
-                compilerArgs.add("--release");
-                compilerArgs.add(targetCompatibilityVersion.getMajorVersion());
-
+                compileTask.getConventionMapping().map("sourceCompatibility", () -> java.getSourceCompatibility().toString());
+                compileTask.getConventionMapping().map("targetCompatibility", () -> java.getTargetCompatibility().toString());
+                compileOptions.getRelease().set(releaseVersionProviderFromCompileTask(project, compileTask));
             });
             // also apply release flag to groovy, which is used in build-tools
             project.getTasks().withType(GroovyCompile.class).configureEach(compileTask -> {
-
                 // TODO: this probably shouldn't apply to groovy at all?
-                // TODO: use native Gradle support for --release when available (cf. https://github.com/gradle/gradle/issues/2510)
-                final JavaVersion targetCompatibilityVersion = JavaVersion.toVersion(compileTask.getTargetCompatibility());
-                final List<String> compilerArgs = compileTask.getOptions().getCompilerArgs();
-                compilerArgs.add("--release");
-                compilerArgs.add(targetCompatibilityVersion.getMajorVersion());
+                compileTask.setSourceCompatibility(compileTask.getSourceCompatibility());
```

Comment:
What is this line doing exactly? Are we just overriding the default convention mapping?

Reply:
oversight. already removed. The --release value is only taken for JavaCompile tasks